Fix total shipping display info when no shipping method is available

tarhi-saad opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
When there is no shipping method, local pickup won't remove the pickup cost from the Cart totals.

The shipping rates don't get updated when there are no shipping options available. So, it will always return the list of the available pickup locations when selecting the "Shipping" option. A solution for this issue will be to check if all available shipping methods are local pickup only. Then when selecting "Shipping", we can determine that there are no shipping options available!

A conflict did arise after merging #8141. With this PR, we display a link to add the shipping address in the shipping calculator on the cart page when the shipping address is incomplete. We had to change the logic implementation for this one to fix the conflict!
<!-- Reference any related issues or PRs here -->

Fixes #8206

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->

| Before | After |
| ------ | ----- |
|    <img width="713" alt="image" src="https://user-images.githubusercontent.com/14235870/228025405-1b779dfc-fc15-47aa-b40e-124781803c71.png">    |     <img width="704" alt="image" src="https://user-images.githubusercontent.com/14235870/228025633-fdffe5a8-aabf-4dd3-bbd4-61c2004a240e.png">  |

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

The order of the test cases is important!

### Case 1 (Local pickup enabled and no shipping options available)

1. Enable local pickup with pickup price. For that, go to `WooCommerce -> Settings -> Shipping -> Local pickup`, under `General` Check `Enable local pickup` and `Add a price for customers who choose local pickup`, and save. Then add a price under the `Cost` input.
2. Remove the shipping methods from the WC settings (WooCommerce -> Settings -> Shipping -> Shipping zones)
3. Go to the store and add an item to the cart
4. Go to the Checkout block and select the `Local Pickup` option under the `Shipping Method` section. Ensure the correct shipping information is displayed in the `Order Summary` Block  (i.e., The correct price and correct local pickup option)
5. Select `Shipping` under the `Shipping Method` section. Ensure `No shipping options available` is displayed under `shipping` in the `Order Summary` Block

### Case 2 (Local pickup enabled and shipping option(s) available)

1. Add a shipping method in `WooCommerce -> Settings -> Shipping -> Shipping zones`. Ensure the shopper's address is within the shipping zone created.
2. Go to the Checkout block and select the `Local Pickup` option under the `Shipping Method` section. Ensure the correct shipping information is displayed in the `Order Summary` Block  (i.e., The correct price and correct local pickup option)
3. Select `Shipping` under the `Shipping Method` section. Ensure the correct shipping information is displayed in the `Order Summary` Block (i.e., The correct price and correct shipping option)

### Case 3 (Local pickup disabled and no default shipping rates)

1. Disable local pickup. For that, go to `WooCommerce -> Settings -> Shipping -> Local pickup`, under `General` uncheck `Enable local pickup` and save.
2. Go to `WooCommerce -> Settings -> General`. Select `No location by default` under `Default customer location`
3. Remove any default(fallback) shipping rates in `WooCommerce -> Settings -> Shipping -> Shipping zones -> Locations not covered by your other zones`
4. Open your site in a private session(incognito mode).
5. Add an item to the cart
6. Go to the cart page, and confirm the "Add an address for shipping rates" link is present in the shipping calculator.
7. Click on the link, and confirm it is working as expected.
8. Open your site in a regular session where you're already logged in and have placed an order before.
9. Add an item to the Cart
10. Go to the cart page, confirm the shipping address, and "change address" link present in the shipping calculator.
11. Click on the link, and confirm it is working as expected.

### Case 4 (With default shipping rates)

1. Add a default(fallback) shipping rate in `WooCommerce -> Settings -> Shipping -> Shipping zones -> Locations not covered by your other zones`
2. Follow steps 4 and 5 from the previous case (i.e., `Case 3`)
3. Go to the cart page, and confirm the shipping calculator is hidden and default shipping rates are visible
4. Follow steps 8 to 11 from the previous case (i.e., `Case 3`)

### Case 5 (Error Notice)

1. Follow steps 3 to 6 from `Case 3`
2. Click on the link, and change the address to a location where shipping rates are unavailable.
3. Confirm that `There are no shipping options available. Please check your shipping address.` error is visible in the text form.
4. Proceed to the Checkout block page, and confirm the `There are no shipping options available. Please check your shipping address` error is visible in the text form under the `Shipping options` section.


Note: Notice will appear for a few seconds until we receive the server's response, even with the valid shipping address, when we change the shipping method. 

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Fix total shipping display info when no shipping method is available

tarhi-saad committed:
Fix total shipping info when no shipping are available

tarhi-saad committed:
Fix a logical error for displaying shipping info

nielslange reviewed (approved):
Thanks for working on this, @tarhi-saad! I've successfully tested the change. While reviewing, I saw that the buyer's notice `No shipping options available` could be improved:

<img width="279" alt="Screenshot 2023-03-28 at 12 54 42" src="https://user-images.githubusercontent.com/3323310/228141480-bbe82999-0863-42d6-9c20-b55f47a0948c.png">

I then recalled that @mikejolley is currently working on [Feature Branch: Updated Buyer Notices](https://github.com/woocommerce/woocommerce-blocks/pull/8659). Once that PR has been merged, I believe we should improve the  `No shipping options available` message in the sidebar. But that's another story. For now, let's ⛴ this PR.

tarhi-saad committed:
Fix failing unit tests

tarhi-saad committed:
Run unit test for the Cart instead of the Checkout

The calculator is only available for the Cart Block, so it doesn't make
sense to run this test for the Checkout Block

tarhi-saad committed:
Fix no shipping methods and incomplete address conflict

When there are no shipping methods (except for local pickup), we would
like to inform the shopper that there are no shipping options available
even though the address is complete

The solution we found is to check the address on the Cart Block only

tarhi-saad committed:
Refactor code

tarhi-saad commented:
Hello @tarunvijwani, @nielslange! We finally fixed a conflict me and @tarunvijwani between this PR and #8141. I added additional testing instructions to ensure we didn't introduce regression! Feel free to merge this PR if you think everything works well!

nielslange commented:
Case 1 / Step 5 states:

> Select `Shipping` under the `Shipping Method` section. Ensure `No shipping options available` is displayed under `shipping` in the `Order Summary` Block

Despite not having any shipping zones defined, I still see `Local Pickup` under `Shipping` in the `Order Summary` Block. 

<img width="2877" alt="Screen Shot 2023-04-04 at 11 21 09" src="https://user-images.githubusercontent.com/3323310/229686241-ce95a588-b10f-4173-9102-88b3127e73c8.png">

> The order of the test cases is important!

Given that the first test case failed, I've stopped the review, as other tests might also fail, given that _"The order of the test cases is important"_.

nielslange reviewed (commented):
See https://github.com/woocommerce/woocommerce-blocks/pull/8819#issuecomment-1495327006

opr commented on assets/js/base/components/cart-checkout/totals/shipping/utils.ts:
```diff
@@ -31,3 +31,31 @@ export const getTotalShippingValue = ( values: {
+			( shippingRatePackage ) =>
+				! shippingRatePackage.shipping_rates.some(
+					( shippingRate ) =>
+						shippingRate.method_id !== 'pickup_location'
```
This will not work for third-party collectible rates, we need to check if the method ID is in `collectableRateIds` from the server.

opr committed:
Check whether rate is collectible without using hardcoded id

opr reviewed (approved):
I was unable to replicate the bug Niels shared.

I have Local Pickup enabled, I have no shipping zones or fallback rates configured. On the Checkout block I never see the local pickup rates unless I specifically select the local pickup button.

I found a bug in the code where the `pickup_location` string was used, however third-party methods supporting `local-pickup` will not be considered in this, I pushed an update to fix this.

We should also add unit tests for that method.

I will merge this.

opr committed:
Correctly negate hasCollectibleRate result

opr commented:
Bumping to 10.1.0 after p1681129771372729/1681128388.500929-slack-C02UBB1EPEF

nielslange reviewed (approved):
I reviewed the PR again and now all testing steps worked as expected. 

@opr I see that both you and @tarhi-saad applied various changes yesterday. Given that @tarhi-saad will be AFK until April 28, shall we merge this PR, if all testing steps are passing for you as well?

tarunvijwani reviewed (changes requested):
I'm little hesitant in approving the PR as this still have shipping total error. 
If I select the shipping and don't enter the valid address, it calculates the local pickup cost but doesn't display it in the order summary. 

| Shipping | Local pickup |
|---|---|
|![image](https://user-images.githubusercontent.com/11503784/232116539-5281d89f-d74b-4a08-ac7b-c587a79ff40e.png)| ![image](https://user-images.githubusercontent.com/11503784/232116668-08a09dd2-1772-4c12-a8ba-8f5b3003e7bf.png)|

opr commented:
The issue here is that if "Shipping" is selected while an address is not entered, there are no "shipping" rates, so it falls back to the local pickup option, which in this case is $4. This is used to calculate taxes, too.

To fix this, we would need to get the taxes and the current shipping price and manipulate the totals in-place to show the correct value. I don't think there's a way we can correct this on the server, because as far as the server is concerned, there is no difference between a local pickup rate and a shippable rate. We would then need to remove the correct amounts from each tax line, if it's being displayed in an itemized list. This is diffficult as we'd need to know what percent each tax line applies, and subtract the shipping tax from each one.

I think it will be _very_ hard to get the correct totals to display when "shipping" is selected but before the address, and with little benefit.

I propose we add a notice saying totals will be calculated when an address is entered.
<img width="1017" alt="image" src="https://user-images.githubusercontent.com/5656702/233792244-e6f6080b-6150-4806-bc82-8d65d58f1e0d.png">

thealexandrelara commented:
WooCommerce Blocks 10.1.0 will be released soon. I'm bumping this PR to the next milestone

opr committed:
Add notice when shipping is selected but no methods are available yet (#9171)

* Create useShippingTotalWarning hook

* Show notices above checkout sidebar

* Call hook to show notice in Checkout block

* Remove unused imports

* Update hook name to useShowShippingTotalWarning

* Move hook to its own file

* Import shipping data internally (without alias)

* Remove unused imports

* Move import to correct place

* Return early to avoid if else

* Refactor useShowShippingTotalWarning

* Get shipping rates directly from the cart instead of the hook

opr commented:
@tarunvijwani please could you re-review this? @nielslange please also feel free to give it another look, it incorporates the changes we made in #9171

nielslange reviewed (approved):
I've successfully tested the PR again, so I pre-approved the PR, but I struggled a bit with the testing instructions, see below:

I wonder if we should refine the testing instructions for better efficiency. The complexity arising from the statement, "The order of the test cases is important!" has made testing more challenging. Each step requires recalling the previous ones. For instance, Steps 2 and 4 of Case 4, as well as Step 1 of Case 5, direct users to revisit steps from prior cases. I found this constant back-and-forth mentally exhausting, reducing my ability to effectively conduct the review. In my view, it would be more efficient if test cases were designed to be independent of one another, allowing for streamlined and focused testing. What are your thought on that, @opr?

tarunvijwani reviewed (commented):
@opr Thank you for adding the notice to the checkout summary. It helps a lot to understand that shipping will be recalculated and avoid confusion. However, I noticed two bugs now:

1. My entered email address get removed automatically when I start entering the shipping address.  
2. Even after adding the correct address, if I select local pickup and select back shipping. It displays the notice, which is not valid anymore. 

Here is the screen recording for your reference: 


https://user-images.githubusercontent.com/11503784/235434644-d8f3aa4c-341d-4572-a9f1-2a7484dd065f.mov

opr commented:
@tarunvijwani thanks, are you able to reproduce the email issue on trunk? I don't see how that could be caused by this PR.

How long did you wait after switching method, it looks like the `select-shipping-rate` request to the server did not complete. This is on the basis that the selected method is still "Pickup" at the end of your screenshare.

<img width="323" alt="image" src="https://user-images.githubusercontent.com/5656702/235436195-25d3cccb-1291-4b25-9329-cd33197b01ba.png">

tarunvijwani commented:
>thanks, are you able to reproduce the email issue on trunk? I don't see how that could be caused by this PR

No, I am unable to reproduce it now. It could be a glitch as I'm not able to reproduce it on this branch as well.

>How long did you wait after switching method, it looks like the select-shipping-rate request to the server did not complete. This is on the basis that the selected method is still "Pickup" at the end of your screenshare.

@opr I tried again, I'm getting the same results even if I wait for a couple of minutes.

opr commented:
Hi @tarunvijwani ok, great well I have not reproduced that and since you can't either I don't think it's anything too problematic, we can keep an eye on this and see if it happens again - the cause will be the server sending a new cart for some reason with the email address missing from the billing address. 

As for the second part, I am not able to reproduce that

https://www.loom.com/share/17fc62458d1b41bcb81686f8d0ab29cb

Would you be able to set up a JN site and have a completely fresh install using the testing zip file?

tarunvijwani commented:
@opr Here is the JN site: https://mean-platypus.jurassic.ninja/checkout/
I can reproduce both issues on it. 
Steps to reproduce:
1. Add a product to the cart. 
2. Go to the checkout page. 
3. Select the shipping option. 
4. Confirm the notice appears in the Checkout summary.  
5. Enter your email address. 
6. Start entering the address. 
7. Notice email address entered is cleared, and a validation error appears. 
8. After entering the address, enter the email address. 
9. Confirm the notice in the checkout summary block no longer appears. 
10. Select local pickup. 
11. Select shipping again. 
12. Notice appears again in the checkout summary block even with the valid address.

tarunvijwani commented:
Created a ticket for email address issue: https://github.com/woocommerce/woocommerce-blocks/issues/9321

opr committed:
Show shipping cost when price information is available

opr committed:
Check if the passed rates are considered selected

kmanijak commented:
PR bumped to the next milestone (10.3.0)

Commit:
Merge branch 'trunk' into fix/total-shipping-display-info

opr committed:
Prevent errors when no rates are available

tarunvijwani reviewed (approved):
Thank you for fixing the errors. I have added a note in the PR description that notice will appear for a few seconds until we receive the server's response, even with the valid shipping address, when we change the shipping method.

Commit:
Merge branch 'trunk' into fix/total-shipping-display-info

mikejolley commented:
Bumping to 10.4.0.

ralucaStan commented:
@tarhi-saad what is preventing this PR from getting merged?

opr commented:
> @tarhi-saad what is preventing this PR from getting merged?

Since I worked on this a lot during Saad's AFK, I asked if he could do one final review/test before merging.

tarhi-saad committed:
Merge branch 'trunk' into fix/total-shipping-display-info

tarhi-saad commented:
Everything works as expected!

> There were two E2E tests that failed in connection to the `Filter by Rating Block`. However, since it has no relation to this block, I will proceed to merge this PR.

It was a flacky test! After running it again it was successful.
